File: khalti-android/src/main/java/khalti/checkOut/EBanking/EBankingPresenter.java
package khalti.checkOut.EBanking;

import android.support.annotation.NonNull;

import java.util.HashMap;
import java.util.List;
import java.util.concurrent.TimeUnit;

import khalti.checkOut.EBanking.helper.BankPojo;
import khalti.checkOut.EBanking.helper.BankingData;
import khalti.checkOut.api.Config;
import khalti.checkOut.api.ErrorAction;
import khalti.utils.EmptyUtil;
import khalti.utils.GuavaUtil;
import khalti.utils.Store;
import rx.Observable;
import rx.Subscriber;
import rx.subscriptions.CompositeSubscription;

public class EBankingPresenter implements EBankingContract.Presenter {
    @NonNull
    private final EBankingContract.View view;
    private EBankingModel eBankingModel;
    private Config config;
    private CompositeSubscription compositeSubscription;

    public EBankingPresenter(@NonNull EBankingContract.View view) {
        this.view = GuavaUtil.checkNotNull(view);
        view.setPresenter(this);
        eBankingModel = new EBankingModel();
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void onCreate(boolean hasNetwork) {
        this.config = Store.getConfig();
        view.toggleIndented(true);
        HashMap<String, Observable<Void>> map = view.setOnClickListener();
        compositeSubscription.add(map.get("try_again").subscribe(aVoid -> onCreate(hasNetwork)));
        if (hasNetwork) {
            compositeSubscription.add(eBankingModel.fetchBankList()
                    .subscribe(new Subscriber<List<BankPojo>>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            view.showIndentedError(e.getMessage());
                            config.getOnCheckOutListener().onError(ErrorAction.FETCH_BANK_LIST.getAction(), e.getMessage());
                        }

                        @Override
                        public void onNext(List<BankPojo> banks) {
                            view.toggleIndented(false);
                            view.setUpList(banks);
                            view.toggleSearch(banks.size() > 3);
                            compositeSubscription.add(view.getItemClickObservable()
                                    .subscribe(hashMap -> view.openMobileForm(new BankingData(hashMap.get("idx"), hashMap.get("name"), hashMap.get("logo"),
                                            hashMap.get("icon"), config))));

                            compositeSubscription.add(view.setSearchListener()
                                    .debounce(500, TimeUnit.MILLISECONDS)
                                    .subscribe(charSequence -> compositeSubscription.add(view.filterList(charSequence + "")
                                            .subscribe(integer -> {
                                                if (EmptyUtil.isNotNull(integer)) {
                                                    view.toggleSearchError(integer == 0);
                                                }
                                            }))));
                        }
                    }));
        } else {
            view.showIndentedNetworkError();
        }
    }

    @Override
    public void onDestroy() {
        if (EmptyUtil.isNotNull(compositeSubscription) && compositeSubscription.hasSubscriptions() && !compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
        eBankingModel.unSubscribe();
    }

    public void injectModel(EBankingModel eBankingModel) {
        this.eBankingModel = eBankingModel;
    }

    public void injectConfig(Config config) {
        this.config = config;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/ebanking/EBankingPresenter.kt
package com.khalti.checkOut.ebanking

import com.khalti.checkOut.api.ErrorAction
import com.khalti.checkOut.api.Result
import com.khalti.checkOut.ebanking.helper.BankingData
import com.khalti.checkOut.helper.Config
import com.khalti.signal.CompositeSignal
import com.khalti.utils.EmptyUtil
import com.khalti.utils.GuavaUtil
import com.khalti.utils.LogUtil
import com.khalti.utils.Store
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class EBankingPresenter(view: EBankingContract.View) : EBankingContract.Presenter {

    private val view: EBankingContract.View = GuavaUtil.checkNotNull(view)
    private var model: EBankingModel = EBankingModel()
    private lateinit var config: Config
    private val compositeSignal = CompositeSignal()

    private val parentJob = Job()
    private val scope = CoroutineScope(Dispatchers.Main + parentJob)

    init {
        view.setPresenter(this)
    }

    override fun onCreate() {
        this.config = Store.getConfig()

        val clickMap = view.setOnClickListener()
        if (EmptyUtil.isNotNull(clickMap["try_again"])) {
            onFetch()
        }
    }

    override fun onFetch() {
        if (view.hasNetwork()) {
            view.toggleIndented(true)
            scope.launch {

                when (val result = model.fetchBankList()) {
                    is Result.Success -> {
                        view.toggleIndented(false)
                        compositeSignal.add(view.setupList(result.data.records)
                                .connect {
                                    view.openMobileForm(BankingData(it.getValue("idx"), it.getValue("name"), it.getValue("logo"), it.getValue("icon"), config))
                                })
                        compositeSignal.add(view.setSearchListener()
                                .connect { t ->
                                    val count = view.filterList(t)
                                    LogUtil.log("i", count)
                                    view.toggleSearchError(count == 0)
                                })
                        view.toggleSearch(result.data.records.size > 3)
                    }
                    is Result.Error -> {
                        val message = result.throwable.message
                        if (EmptyUtil.isNotNull(message)) {
                            view.showIndentedError(message!!)
//                            config.onCheckOutListener.onError(ErrorAction.FETCH_BANK_LIST.action, message)//TODO
                        }
                    }
                }
            }

        } else {
            view.showIndentedNetworkError()
        }
    }

    override fun onDestroy() {
        compositeSignal.clear()
        parentJob.cancel()
    }
}
